keps/sig-network/2086-service-internal-traffic-policy/README.md

```diff
@@ -115,56 +112,68 @@ Proposed addition to core v1 API:
-	ServiceInternalTrafficPolicyTypeLocal       ServiceInternalTrafficPolicyType = "Local"
+	ServiceTrafficPolicyTypeCluster     ServiceTrafficPolicyType = "Cluster"
+	ServiceTrafficPolicyTypeTopology    ServiceTrafficPolicyType = "Topology"
+	ServiceTrafficPolicyTypePreferLocal ServiceTrafficPolicyType = "PreferLocal"
+	ServiceTrafficPolicyTypeLocal       ServiceTrafficPolicyType = "Local"
 )
 
 // ServiceSpec describes the attributes that a user creates on a service.
 type ServiceSpec struct {
 	...
 	...
 
-	// internalTrafficPolicy denotes if the internal traffic for a Service should route
-	// to cluster-wide endpoints or node-local endpoints. "Cluster" routes internal traffic
-	// to a Service to all cluster-wide endpoints. "PreferLocal" will route internal traffic
-	// to node-local endpoints if one exists, otherwise it will fallback to the same behavior
-	// as "Cluster". "Local" routes traffic to node-local endpoints only, traffic is dropped
-	// if no node-local endpoints are ready.
-	InternalTrafficPolicy ServiceInternalTrafficPolicyType `json:"internalTrafficPolicy,omitempty"`
+	// trafficPolicy denotes if the traffic for a Service should route
+	// to cluster-wide endpoints or node-local endpoints. "Cluster" routes traffic
+	// to a Service to all cluster-wide endpoints. "Topology" routes traffic based on
+	// topology hints. "PreferLocal" will route internal traffic to node-local endpoints
+	// if one exists, otherwise it will fallback to the same behavior as "Cluster".
+	// "Local" routes traffic to node-local endpoints only, traffic is dropped
+	// if no node-local endpoints are ready. When externalTrafficPolicy is "Cluster",
+	// traffic from external sources will be routed based on the trafficPolicy. When
+	// externalTrafficPolicy is "Local", trafficPolicy is ignored for traffic from
+	// external sources.
+	// +optional
+	// +feature-gate=ServiceTrafficPolicy
+	TrafficPolicy ServiceTrafficPolicyType `json:"trafficPolicy,omitempty"`
 }
 ```
 
+This new field will intersect with externalTrafficPolicy in the following ways:
+* if `externalTrafficPolicy=Cluster`, traffic will be routed based on `trafficPolicy` for external sources
+* if `externalTrafficPolicy=Local`, `externalTrafficPolicy` will take precedent over `trafficPolicy`, but only for external sources.
+
 Proposed changes to kube-proxy:
-* when `internalTrafficPolicy=Cluster`, default to existing behavior today.
-* when `internalTrafficPolicy=PreferLocal`, route to endpoints in EndpointSlice that matches the local node's topology (topology defined by `kubernetes.io/hostname`),
+* when `trafficPolicy=Cluster`, default to existing behavior today.
+* when `trafficPolicy=Topology`, use topology hints from EndpointSlice API.
+* when `trafficPolicy=PreferLocal`, route to endpoints in EndpointSlice that matches the local node's topology (topology defined by `kubernetes.io/hostname`),
 fall back to "Cluster" behavior if there are no local endpoints.
```

Comment:
We'll have to figure out if the fallback is Cluster or Topology or if we need 2 modes.  OK for here, the PR will need to explore that.